### ISSUE TITLE
feat: add an abstract layer to support pydantic/msgspec

### ISSUE THREAD

**Issue by kemingy** [0b01001001/spectree#329]
Pydantic v2 has too many breaking changes that lots of 3rd libraries are not able to upgrade in a short time.

This PR is not backward compatible (due to pydantic breaking changes). I'm not sure if we should make this library compatible with both pydantic v1 & v2. Or maybe we should add another abstract layer so that we can make it possible to support pydantic v1 & v2 and msgspec (ref #307)

_Originally posted by @kemingy in https://github.com/0b01001001/spectree/pull/319#pullrequestreview-1510211443_

**Comment by BradLucky** [comment]
Is this issue/repo still getting attention? This is over a year old at this point. I managed to copy `GenericModel` from Pydantic, making it so that SpecTree can still be used to depend on Pydantic V1 even though we are running V2. SpecTree has been great for us, but if it will not be getting true support for V2, I'm afraid I'll have to remove the dependency and find another solution.

**Comment by kemingy** [comment]
#387 makes it possible to support both pydantic v1 & v2.

**Comment by alexted** [comment]
@kemingy Thank you so much for taking the time and making the effort to look into this matter

**Comment by alexted** [comment]
Hi! I understand you've decided to break the work down into several stages, two of which have already been completed. 
Are there still a few more steps to go?

**Comment by kemingy** [comment]
> Hi! I understand you've decided to break the work down into several stages, two of which have already been completed. Are there still a few more steps to go?

Yeah, the next step is to introduce `msgspec` to see if there are any missing parts need to be abstracted. I'm already working on this one.

**Comment by kemingy** [comment]
Hi @alexted, the PR #479 can pass some basic msgspec tests. Can you give it a try? Looking forward to your feedback!

**Comment by alexted** [comment]
Sorry for the delayed response, it took some time. Thank you for your work - it looks great and works like magic! 
The adapter is a solid foundation. I left you a detailed comment in the PR.

**Comment by alexted** [comment]
@kemingy  

- Could you clarify whether it will now be possible to create custom adapters for other serializers/validators, or if the current adapter abstraction doesn’t support that?

- This new functionality seems like it would warrant a new major version, doesn’t it, huh?

**Comment by kemingy** [comment]
> [@kemingy](https://github.com/kemingy) Could you clarify whether it will now be possible to create custom adapters for other serializers/validators, or if the current adapter abstraction doesn’t support that? This new functionality seems like it would warrant a new major version, doesn’t it, huh?

- it supports custom model adapter, you can check the model adapter protocol, as long as the adapter can implement its validation and OpenAPI schema generation
- yes, I think I have to create a new major version. I was thinking the interface should almost unchanged, but it actually changed a bit. So a major version is necessary.

**Comment by kemingy** [comment]
The alpha https://github.com/0b01001001/spectree/releases/tag/v3.0.0a0 has been released.

Please give it a try.

**Comment by alexted** [comment]
Thanks, it works like magic!
